### PR TITLE
HPCC-26254 Generate numDiskReads for keyed join

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Review by jakesmith — CHANGES_REQUESTED** [review]
@shamser - there's an issue with tracking all active index parts.

Also, this doesn't deal with the remote side of KJ, i.e. where the *Remote*Handler's are involved. However, I think I'd leave that to do in a separate PR, as it's significantly different.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1840]: related to previous comments, could do once in CLookupHandler::init()

File: fs/dafsclient/rmtfile.cpp
Comment (jakesmith) [inline fs/dafsclient/rmtfile.cpp:1353]: looks like a correct addition/change, but not directly related to this JIRA?

File: system/jhtree/jhtree.cpp
Comment (jakesmith) [inline system/jhtree/jhtree.cpp:2307]: think you should change this to:
```
    virtual void mergeStats(CRuntimeStatisticCollection & stats) const override { if (!realKey) return; realKey->mergeStats(stats); }
```

because you don't want to force the key to open when it hasn't been just because asking for stats. (for which there won't be any yet).

**Review by jakesmith — CHANGES_REQUESTED** [review]
@shamser - the structure looks good. I think need to address the potential thread safety issues on the calls interacting with the key manager, and would be good to move the mapping so not done each call.

File: fs/dafsclient/rmtfile.cpp
Comment (shamser) [reply fs/dafsclient/rmtfile.cpp:1353]: The disk join is using the diskReadActivityStatistics and this statistics mapping  has StTimeLocalExecute in its list of stats.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1378]: @ghalliday reckons reading an atomic will be no slower, so a separate boolean vector is not worth it, may as well make the keyManagers objects atomics.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1371]: parts is an array of unsigned:     std::vector<unsigned> parts

The only parts descriptor that I see is allParts.
@jakesmith

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1380]: Shall I use the top 8 bits of partNo to store the subfile number?

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1419]: Please see previous comment relating to this (https://github.com/hpcc-systems/HPCC-Platform/pull/15278/files#r702905024)

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1838]: Shall I be using queryFetchPartIO ?  I.e. Should it be opening the stream if it's not already open? 

Can you also clarify why the atomic is needed?  Can the partIO.iFileIO be closed by another thread?

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1371]: right my mistake, allParts was intended for tracing only (only use of it was in trace() which is in _DEBUG).
However, you could remove the comment that says that, and use it for this purpose,
or you could use the underlying parts array in each case, either activity.allDataParts

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1380]: I think I'd store them in a vector, such that you can lookup with partSubFile[<partNo>]

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1838]: >Shall I be using queryFetchPartIO ? I

no, because that would created it, and there's no need for this thread to, there will be no stats. if not created.

>Can you also clarify why the atomic is needed? Can the partIO.iFileIO be closed by another thread?

it can't be closed, but it's not strictly thread safe for 1 thread to write a pointer, whilst another concurrently reads it.
In reality on Intel architectures it will be fine, but not necessarily on all, on some it could write 1st 32bit of pointer, and the reading thread uses that before fully initialized.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1371]: I'll remove the comment and use allParts because it will allow init() to setup the subfile vector for both index and data files.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1423]: Am I correct in understanding that process() does not have to be thread safe?

Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1423]: It was thread safe, as there was only 1 thread in process() (Each handler has 1 process() thread,) and all the entities it dealt with were either it's own, pre-intiialized, or made thread safe on access.
The introduction of another thread that interacted with things the process() thread initializes meant there was potential race conditions that needed to be protected from.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1436]: >N.B. This isn't ensure previous line is not executed multiple times.

that bit of the comment isn't correct, the keyManager pointer (line 1425) was/is sufficient for that - this is the only thread that will touch it. This boolean is only to prevent other threads (i.e. the thread that's collecting the stats) from accessing the keyManager before it's created.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1542]: trivial: could add override

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1778]: this change isn't needed. The process() thread is the only thread that updates partIOs

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1800]: I think these atomic need initializing.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1868]: this should move inside the conditon below,
Should also change so it refernces the PartIO object, it's currently copying it (which involves linking all it's members etc).

**Review by jakesmith — CHANGES_REQUESTED** [review]
@shamser - a few comments on new commit

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1371]: I did check and they were initialized to false.  However,  if the standard doesn't specify that it's guaranteed to initialized to false,  I'll modify the code.

Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1371]: I'm still not sure reading the docs, but we initialize atomics elsewhere, and probably better to be explicit anyway.
```
keyManagersCreated.reset(new std::vector<std::atomic_bool>(parts.size(), false));
```

Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1369]: That doesn't work because std::atomic does not have a copy constructor.  I'm using a loop to initialize.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1874]: still copying the object, should be a reference:
PartIO &partIO = partIOs[i];

**Review by jakesmith — COMMENTED** [review]
@shamser - just 1 issue.

**Review by jakesmith — COMMENTED** [review]
@shamser - looks good, please squash

**Review by jakesmith — APPROVED** [review]
@shamser - looks good

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:2443]: I don't think this is the correct statistics mapping

**Review by ghalliday — CHANGES_REQUESTED** [review]
@shamser various comments about atomics and which statistic mappings are being used.

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoin.cpp:314]: It shouldn't, there are bigger problems if there 100's of subs in supers..
a reserve would be no harm, but I think inconsequential with everything else going on in this init() function (like calling Dali etc.)

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1352]: it's not possible to size a vector of atomic's because they have no move or copy ctor.
So you can only make a const vector size, and so need to allocate if using std::vector of atomics afaics.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1399]: @ghalliday - that's something I missed, I thought there'd be a barrier with a standard memory order, not just effecting or of other atomic operations. Is there no way to ensure that atomic test is not reordered at all?

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1417]: In this case, it would be simple to make the IKeyManager an atomic instead.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (ghalliday) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1352]: ok, That's worthy of a comment, because not obvious,
In future could also use
```
std::unique_ptr<std::atomic_bool []> x;
x.reset(new std::atomic<bool>[n]);
```

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (ghalliday) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1399]: I was wrong.  The atomic write will guarantee that all previous writes are committed, and the read will prevent any reads being migrated before it.
It might be worth using explicit store/load memory order parameters, but in this case the inefficiency of sequential ordering  is unlikely to matter (and doesn't really affect intel).

**Review by ghalliday — COMMENTED** [review]
@shamser please disregard my comments about the atomics.  Once the other comments are addressed I'll re-review.

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoin.cpp:385]: I'll simplify.

The change to numSuperIndexSubs will also affect the totalIndexParts calculation in line 414.     Will that cause any issues?

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoin.cpp:577]: I've examined the numDiskReads count for the index file and index file's subfiles and have found that the recorded numDiskReads change everytime I execute a keyed_joined query.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:2433]: Just consistency.  Couple of lines down it is using lazyIFileIO.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (shamser) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1774]: Do you mean getting rid of partIOCreated and using unique_ptr<std::vector<std::atomic<std::unique_ptr<partIOs>>>> partIOs?

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1399]: phew, ok that's good to confirm, it sounded that way form my scanning of the docs (a short while ago).
Prior to that I assumed the standard memory ordering of atomics guaranteed it didn't move around reads..
If it did, I suspect (and we were on compilers/architecture where it mattered), my guess was that we'd have other similar code that was affected

File: system/jhtree/jhtree.cpp
Comment (jakesmith) [reply system/jhtree/jhtree.cpp:2311]: doesn't need to call checkOpen() if realKey is not null, it is open (there's nothing in the impl. which will release it).
So could be:
```
    {
        CriticalBlock b(c);
        if (!realKey) return;
    }
    realKey->mergeStats(stats);
```

File: thorlcr/activities/keyedjoin/thkeyedjoin.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoin.cpp:385]: that line is fine, keyHasTlk will only be true if it's been through the loop, i.e. if there is actually any subs.

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1387]: Was there an existing bug where process could be called on multiple threads?  If not, is the criticalBlock needed?

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1781]: Does this need to change?  The previous code looks fine to me.  Is this trying to protect partIO.stream?

**Review by ghalliday — COMMENTED** [review]
@shamser looks close to ready.  A couple of questions

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1387]: There is only 1 process() thread for this object and it's the only thread that updates the keyManagers array, so shouldn't need CS/double check.

I think as before, you should take the vector item by reference and assign, except now it's an atomic. e.g.:
```
            std::atomic<IKeyManager *> &keyManager = (*keyManagers)[selected];
            if (!keyManager) // delayed until actually needed
            {
                keyManager = activity.createPartKeyManager(partNo, copy);
```

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1781]: Atomic on a PartIO pointer don't really help.
queryFetchPartIO is overwriting the PartIO objects (not pointers), the pointers never change after init in this version.

I think the was correct before this latest commit, using an atomic partIOCreated, which is checked by the stats call.
(since we clarified that the atomic read code is effectively fenced).

**Review by jakesmith — COMMENTED** [review]
@shamser - see comments

**Review by jakesmith — CHANGES_REQUESTED** [review]
updating as review status wrong (needed to add any comment)

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (jakesmith) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1799]: This code needs to be removed / is causing the compile error.

**Review by jakesmith — CHANGES_REQUESTED** [review]
compile issue

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1353]: I think this can now be deleted.

**Review by ghalliday — COMMENTED** [review]
@shamser Looks ready.  One thing to be deleted, and will need squashing once Jake has approved.

**Review by jakesmith — COMMENTED** [review]
Looks good apart from the removal of the CS @ghalliday already commented on.
Please remove and squash.

**Review by jakesmith — APPROVED** [review]
@shamser - didn't notice this was ready (github review status not refreshed after last push), 
All looks good.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**
